Fix reading of empty S3 files

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix reading of empty S3 files

It is possible for ReadBufferFromS3::nextImpl() called even after eof(),
at least once, and in this case, if the file was empty, then local
working_buffer will be null, while impl.working_buffer will be empty,
but not null, and so local position() after impl->position() =
position() will be incorrect.

I found this with test_storage_s3/test.py::test_empty_file in debug
build, assertion catched this, so maybe it worth get back debug
integration build...

azat committed:
Fix reading of empty S3 files

It is possible for ReadBufferFromS3::nextImpl() called even after eof(),
at least once, and in this case, if the file was empty, then local
working_buffer will be null, while impl.working_buffer will be empty,
but not null, and so local position() after impl->position() =
position() will be incorrect.

I found this with test_storage_s3/test.py::test_empty_file in debug
build, assertion catched this, so maybe it worth get back debug
integration build...

v2: fix test_log_family_s3 failures
    https://s3.amazonaws.com/clickhouse-test-reports/37801/b5e6e2ddae94d6a7eac551309cb67003dff97df1/integration_tests__asan__actions__[2/3].html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

kssenii commented:
you mean this assertion `assert(!internal_buffer.empty());` in line 117?

kssenii commented on src/IO/ReadBufferFromS3.cpp:
```diff
@@ -152,11 +152,12 @@ bool ReadBufferFromS3::nextImpl()
     if (!next_result)
         return false;
 
-    BufferBase::set(impl->buffer().begin(), impl->buffer().size(), impl->offset()); /// use the buffer returned by `impl`
```
I do not see how this should fix failed assertion. Why `hasPendingData` will become `true` if we reached end of file and so size() and offset() should be 0?

azat replied:
In case of empty file `BufferBase::set()` will never be called.
So when it first enters `ReadBufferFromS3::nextImpl()` everything is correct it returns false.
However when it will be called next time, it will enter here https://github.com/ClickHouse/ClickHouse/blob/0b40e05ffcb9e5986425da0d6f61969894168017/src/IO/ReadBufferFromS3.cpp#L98-L99

Before those lines `impl->position()` is `impl->working_buffer.end_pos`,
after assignment `impl->position()` will be corrupted, since it will `position()` is just NULL,
and so `impl->hasPendingData()` (which is `available() > 0`), will return some huge number (non-null pointer - NULL)

kssenii commented:
though failed integration tests are related `Code: 23. DB::Exception: Received from 172.16.14.6:9000. DB::Exception: Unexpected state of istream at offset 0`. This should be related to what is done under `use_external_buffer`. Not sure, but my guess is that the reason in this line or close https://github.com/ClickHouse/ClickHouse/blob/master/src/Disks/IO/ReadBufferFromRemoteFSGather.cpp#L250. May be a  check for `if (result)` is just missing.

azat committed:
tests: improve test_log_family_s3 in case of failures

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: use id for parametrized tests in test_log_family_s3

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: test_log_family_s3 apply black

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix assertion for getImplementationBufferOffset() for Log family on S3

Test test_log_family_s3/test.py::test_log_family_s3[TinyLog]:

    clickhouse: ./src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:213: virtual bool DB::AsynchronousReadIndirectBufferFromRemoteFS::nextImpl(): Assertion `file_offset_of_buffer_end == impl->getImplementationBufferOffset()' failed.

v2: fix assertion instead of adjusting file_offset_of_buffer_end in ReadBufferFromRemoteFSGather.cpp
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

kssenii commented on src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:
```diff
@@ -240,15 +240,7 @@ bool ReadBufferFromRemoteFSGather::readImpl()
-        /// For log family engines there are multiple s3 files for the same clickhouse file
-        file_offset_of_buffer_end += current_buf->available();
-    }
+    file_offset_of_buffer_end = current_buf->getFileOffsetOfBufferEnd();
```
Why after this change it works correctly if in this case `file_offset_of_buffer_end` will equal file offset of buffer end of current buffer, which is one of the multiple files in s3, but in fact here `file_offset_of_buffer_end` needs to be an accumulated offset among all files as this class acts like a combinator for multiple files in s3 which refer to a single file in ch.

azat replied:
>Why after this change it works correctly if in this case file_offset_of_buffer_end will equal file offset of buffer end of current buffer, which is one of the multiple files in s3, but in fact here file_offset_of_buffer_end needs to be an accumulated offset among all files as this class acts like a combinator for multiple files in s3 which refer to a single file in ch.

I guess not enough tests.
But I guess you are right, instead of adjusting `file_offset_of_buffer_end` assertion should be fixed.

azat commented:
>I found this with test_storage_s3/test.py::test_empty_file in debug
>build, assertion catched this, so maybe it worth get back debug
>integration build...

By the way, what do you (and others) think about getting back debug builds for integration tests?
@alesapin seems that they got lost after converting to github actions.

kssenii commented:
> By the way, what do you (and others) think about getting back debug builds for integration tests?

Of course it would be good, but as I remember, they were turned on in CI in always green mode and failed with a limit of failed queries almost immediately (and half of the times due to "harmful" functions in third party libraries) and fixing of them did not proceed for some time so they were removed, or may be indeed they were just not transferred during CI migrations.

kssenii commented:
Integration tests -- nothing related to s3.

ClickHouse build check (actions) — 21/22 artifact groups are OK (7 of 8 builds are OK)
<img width="1010" alt="Screenshot 2022-06-07 at 23 34 03" src="https://user-images.githubusercontent.com/54203879/172486700-4614b356-b6d5-47f2-bd35-c0614d18c6c6.png">
ClickHouse special build check (actions) — 3/7 artifact groups are OK (3 of 7 builds are OK)
<img width="1156" alt="Screenshot 2022-06-07 at 23 34 20" src="https://user-images.githubusercontent.com/54203879/172486753-5dbfc298-bdf1-400e-9e2b-de7671d74a8e.png">
